### ISSUE TITLE
Alternative ingredients

### ISSUE THREAD

**Issue by blabug** [tstehr/RecipeMD#49]
Got this idea from a pizza recipe, where it specified multiple valid Flour types.
While #4 could be used to express this thing, I think that it is a bit clumsy to express such xor Relations for single ingredients.

My suggestion:
```markdown
- *1kg* Weizenmehl Type 405
  or *1kg* Weizenmehl Type 550
```

For the spec, we consider it an alternative ingredient, if it comes after a linebreak and contains a valid amount. The text before the amout is ignored.

I know that this won't work with ingredients without amount, but it was the best idea I could think of.

Also useful for fresh and dried yeast (different amounts/amount units)

**Comment by tstehr** [comment]
> Got this idea from a pizza recipe, where it specified multiple valid Flour types.
> 
> While #4 could be used to express this thing, I think that it is a bit clumsy to express such xor Relations for single ingredients.

Agreed, that seems like overkill and would lead to variants that basically restate the ingredient name 

> My suggestion:
> ```markdown
> - *1kg* Weizenmehl Type 405
>   or *1kg* Weizenmehl Type 550
> ```

For this particular case I'd just create once ingredient with a compound description, since the amount is the same for both types of flour:

```md
- *1 kg* Weizenmehl Type 405 oder Weizenmehl Type 550
```

> Also useful for fresh and dried yeast (different amounts/amount units)

The situation where the different ingredients use different amounts seems like the stronger justification for a feature like this
to me.

I did a quick search for cases like this and found quite a few of them ([1], [2], [3], [4], [5], [6]). Number [2] is exactly the
case yeast case you mentioned, btw :)

[1]: https://github.com/AberDerBart/recipes/blob/master/general_tsos_chickpeas.md?plain=1#L29
[2]: https://github.com/haeckse/recipes/blob/main/bagels.md?plain=1#L12
[3]: https://github.com/tstehr/recipes/blob/master/Creamy_Pumpkin_Mac_n_Cheese.md?plain=1#L13
[4]: https://github.com/tstehr/recipes/blob/master/Der_beste_Kaesesalat.md?plain=1#L13
[5]: https://github.com/tstehr/recipes/blob/master/Veganes_Coffee_Kakao_Bier_Chili.md?plain=1#L20
[6]: https://github.com/rixx/recipes.rixx.de/blob/main/ausprobieren/vegane-linzer-torte.md?plain=1#L9

> For the spec, we consider it an alternative ingredient, if it comes after a linebreak and contains a valid amount. The text before the amout is ignored.

We would still need to store the text before the amount somehow, otherwise it will get lost during a parse/serialize cycle. Not sure where to put it, maybe a sort of `ingredientPrefix` on the ingredient?

I'm also a little concerned with putting that much semantics into a line break. People tend to have opinions on line length and often hard wrap at that point, which would change the meaning for alternative ingredients in this proposal.

> I know that this won't work with ingredients without amount, but it was the best idea I could think of. 

That would be okay with me. As said above the use case becomes a lot weaker if the amount for both ingredients is the same.

**Comment by tstehr** [comment]
An alternative approach here would be to wait until we get a decision on #1 and then extend that to also cover the ingredient name (and possibly other text in the recipe as well).

That would also allow ingredients like `*2 tbsp* tomato paste (*40 g*)` to work correctly.

**Comment by tstehr** [comment]
> I don't see this as such a big problem […]

Let me expound on my concerns for a bit. The main thing going on is that the parser interprets the newline as the alternative ingredient separator, while any casual user will see the "or" (or another string) and consider that the separator. Differing interpretations make it very easy for incorrect recipes to sneak in. Some examples:

- These ingredients would parse differently, even though they look very similar to a user and in fact render to exactly the same HTML.

  ```markdown
  - *1kg* Weizenmehl Type 405 or *1kg* Weizenmehl Type 550
  - *1kg* Weizenmehl Type 405
  or *1kg* Weizenmehl Type 550
  - *1kg* spring onions, white part only, thinly sliced or *1kg* shallots, peeled
     and cut into quarter rings
  - *1kg* spring onions, white part only, thinly 
     sliced or *1kg* shallots, peeled and cut 
  into quarter rings
  - *1kg* spring onions, white part only, thinly sliced 
     or *1kg* shallots, peeled and cut into quarter rings
  ```

- Using a different prefix could create very confusing ingredients where casual interpretation will not match the parse result.

  ```markdown
  - *200g* mozarella 
    and *100g* parmesan
  - *200g* onions 
    and *100g* shallots
    or *300g* onions
  ```

**Comment by tstehr** [comment]
What do you think about using a slash as the separator for alternative ingredients? That would match common use of the slash and not have the `prefix` problem. Example:

```markdown
- *1kg* Weizenmehl Type 405 / *1kg* Weizenmehl Type 550
```

One open question here is if we could still allow slashes in the ingredient name without that creating an alternative ingredient. We could also apply the heuristic of "does it contain an valid amount" here, but I'd want to think about the consequences of this in more detail before committing to it.

**Comment by tstehr** [comment]
> > What do you think about using a slash as the separator for alternative ingredients?
>
> I don't have a problem doing it this way 👍

I think I prefer slashes over line breaks + prefix, if you're also find with it let's pursue slashes for now and see where we get to.

> I wouldn't allow slashes in the ingredient at all! It will make confusing things like *1kg* Weizenmehl Type 405/550, which provide no parsing benefit over *1kg* Weizenmehl Type 405 or Type 550

I don't think that's going to work. I just did a very cursory search and found all of the following. These all seem like perfectly reasonable ingredients to me which would be torn apart if the slash in them was to be interpreted as an alternative ingredient marker:

```
fresh mozzarella, cut into 1/2-inch cubes (optional)
firm tofu (natural/"Naturtofu")
flavourful vegetable broth/stock, divided
grüne oder rote Chilis (mehr/weniger nach Geschmack)
Gelbe/orangene Marmelade
liquid aminos (or tamari/soy sauce)
Milchersatz, vegan (Reis-/Dinkel-/Hafer-/Sojamilch)
Paprika (rot/gelb)
russet potatoes, peeled and sliced 1/8 inch thick on a mandoline slicer (5 to 6 medium potatoes; see note)
schwarze und/oder grüne Oliven
Shiratamako (glutinous rice flour/sweet rice flour)
Wadschunken (Hesse/Beinscheibe) vom Rind
Weichkäse (Brie/Camembert)
zum Servieren: Pinienkerne, geröstet und/oder Parmesan
```

For all of these, a rule as follows would work: A slash is interpreted as an alternative ingredient divider iff it is followed by an emphasis with only whitespace in between.
